### ISSUE TITLE
Watcher could detect stalled syncing of itself and the underlying Ethereum node

### ISSUE THREAD

**Issue by pdobacz** [omisego/elixir-omg#814]
### Motivation

There were cases when a stalled Watcher went unnoticed. Not only not is that a nuisance it also is a security issue; if the Watcher is stalled it will miss any byzantine event requiring action.

We can't assume that the consumers of Watcher's API will know how to and be able to monitor the results of `status.get` to ascertain the Watcher is not stalled.

### Suggested solution

add an item to `byzantine_events` returned from `status.get`:

```
{
  event: "stalled_syncing",
  details: {
    service: "eth_node" | "block_getter" | "depositor" | ...,
    <possibly some other useful metadata>
  }
}
```

The height taken into account would always be that of the root chain i.e. Ethereum, which is the common point of reference (i.e. not the child chain's `blknum`, which is kinda secondary).

### Behavior proposal

For every service from `services_synced_heights`, calculate the speed at which it progresses through Ethereum. Using this speed calculate the `estimated_time_of_sync(service)`, being when the services is fully synced to a **recent Ethereum height**. "Recent" here means recent wrt. the local datetime! Also this calculation must take the `finality_margin` behavior into account - for some services being synced means lagging several blocks behind recent.

If the `estimated_time_of_sync(service)` is too far off in the future, emit event.

(**NOTE** that just detecting that `synced_height` doesn't change is not enough - we want to detect dangerously slow syncing too!)

### Discussion points

1. Relying on metrics monitoring was suggested by @kasima. I'd opt to elevate it from monitoring/alerting to a feature of the API, as it is a critical security item. Also for the sake of users who don't monitor. And also to keep the notion of `byzantine_events == [] <=> all OK`. This way the user has only a single place to check for the security-critical events.
2. This might obsolete the name `byzantine_events`. Possible replacements: `protocol_alerts`, `actionable_protocol_events`.

**Comment by kevsul** [comment]
I'm in favour of this.
It is a critical security issue, but currently a normal Watcher user would miss it.

It think `status.get` (or push events) is the right place for it too. Right now the Watcher only reports issues with the chain/protocol, but it should also be able to report any issues with itself, as those issues almost certainly affect its ability to report the former.

**Comment by InoMurko** [comment]
Could we - in the same breath - expose _alarms as json_ on a separate endpoint?

**Comment by pdobacz** [comment]
> Could we - in the same breath - expose alarms as json on a separate endpoint?

I like the idea, but maybe not to be delivered "in the same breath". These two feel fairly orthogonal in terms of implementation and take a different perspective. Also these efforts could be perfectly made in parallel.

This issue here is to provide sth like an "end result answer" to the question "is Watcher watching?" (basically "is the Watcher observing recent state of the world?"). The alarms endpoint/section is much more low-level/detailed/poweruser-oriented, something to shed light on why is the Watcher not behaving properly.

**Comment by InoMurko** [comment]
YES

**Comment by pdobacz** [comment]
A side note coming from @paulperegud and myself figuring out watcher specs.

We had this thought that this could potentially be a strictly better mechanism than the current `block_withholding` detection. We were pondering this, but we were not sure. 

(in short, current withholding detection does a bunch of retries to `child_chain/block.get` and if it fails for longer than a threshold it barfs and reports a byzantine event)

I think this topic could get revisited, after the final `stalled_syncing` behavior gets fleshed out (or even implemented).

Also note, it might be the case the current "low-level" withholding mechanism is worthwhile anyway, or covers some scenarios that `stalled_syncing` wouldn't that well, so let's not remove prematurely.

**Comment by InoMurko** [comment]
```
 if it fails for longer than a threshold
```
it's good to be aware of this threshold when it comes to deploying and maintaining child chain servers. cc @jbunce @kendricktan 
this means that Childchain availability SLA is bound by watcher's byzantine detection.

**Comment by unnawut** [comment]
Closing in favor of these issues to have more specific discussions:

- Ethereum height sync stalling -> #1283
- Byzantine events clarification -> #811 
- Watcher internal stall detection -> #1313
